### ISSUE TITLE
Lizmap 3.1.11 / QGISServer 2.18 : info popup is ok but clicking on edit button opens the form for another feature

### ISSUE THREAD

**Issue by mathieubossaert** [3liz/lizmap-web-client#862]
Hi, here is a bug  I solved by downgrading from QGIS Server 2.18 to 2.14 : 

When I clik on a feature, I can see the good information about the feature, when i want to edit it, it opens the form for another feature (with a wrong id).

**Comment by rldhont** [comment]
It's a QGIS Server FeatureID issue. If the primary key is not an integer and only an integer, featureID can't be used has an identifier. So we need to fix QGIS Server and Lizmap.

**Comment by rldhont** [comment]
@mathieubossaert can you provide come screenshot or more information about the primary key and the feature id:
* screenshot of the fields tab of the layer properties window
* screenshot of the popup (see the primary key value)
* screenshot of the edition form (see the primary key value)
* the url generated by lizmap to get the edition form (with your browser web tools)

Or can you give us a link to test ?

**Comment by rldhont** [comment]
@mathieubossaert is`"gid"` defined as the primary key in the datasource of the layer ? Can you take a look at the `path` in the General Tab ?

**Comment by mathieubossaert** [comment]
Here is the datasource : 
`dbname='ll' host=www.xxx.yyy.zzz port=5432 user='toto' sslmode=disable key='gid' srid=2154 type=MultiPolygon table="mdo"."affiche_repro_site" (geometrie) sql=`

mdo.affiche_repro_site is a view.

**Comment by rldhont** [comment]
@mathieubossaert can you test the WFS GetFeature request with the featureID ?
http://localhost:8082/websig/lizmap/www/index.php/lizmap/service/?project=mdo_admin&repository=mdo&SERVICE=WFS&VERSION=1.0.0&REQUEST=GetFature&TYPENAME=affiche_repro_site&featureId=affiche_repro_site.446

**Comment by mathieubossaert** [comment]
```xml
<wfs:FeatureCollection xsi:schemaLocation="http://www.opengis.net/wfs http://schemas.opengis.net/wfs/1.0.0/wfs.xsd http://www.qgis.org/gml http://qgiserver/cgi-bin/qgis_mapserv.fcgi?map=%2Fhome%2Fmdo%2Fmdo_admin.qgs&SERVICE=WFS&VERSION=1.0.0&REQUEST=DescribeFeatureType&TYPENAME=affiche_repro_site&OUTPUTFORMAT=XMLSCHEMA"><gml:boundedBy><gml:Box><gml:coordinates cs="," ts=" ">0,0 0,0</gml:coordinates></gml:Box></gml:boundedBy></wfs:FeatureCollection>
```

**Comment by rldhont** [comment]
Can you test ? http://localhost:8082/websig/lizmap/www/index.php/lizmap/service/?project=mdo_admin&repository=mdo&SERVICE=WFS&VERSION=1.0.0&REQUEST=GetFature&TYPENAME=affiche_repro_site
And search for 446 ?

**Comment by mathieubossaert** [comment]
Result is empty.
When I ask for the getcapabilities, I can see that the layer is referenced as : 
```xml
<FeatureType>
   <Name>Sites_de_repro</Name>
   <Title>Sites de repro</Title>
   <Abstract/>
     <SRS>EPSG:2154</SRS>
     <Operations><Query/><Update/></Operations>
     <LatLongBoundingBox maxx="859320" maxy="6.51597e+06" minx="664145" miny="6.16879e+06"/>
</FeatureType>
```

**Comment by mathieubossaert** [comment]
So if I ask for  : http://localhost:8082/websig/lizmap/www/index.php/lizmap/service/?project=mdo_admin&repository=mdo&SERVICE=WFS&VERSION=1.0.0&REQUEST=GetFeature&TYPENAME=Sites_de_repro
I get thi response for site n 446 : 
```xml
<gml:featureMember>
	<qgs:Sites_de_repro fid="Sites_de_repro.446">
		<gml:boundedBy>
			<gml:Box srsName="EPSG:2154">
			<gml:coordinates cs="," ts=" ">767375.15697262,6264543.50889878 767676.51028226,6264842.25259768
			</gml:coordinates>
			</gml:Box>
		</gml:boundedBy>
		<qgs:geometry>
			<MultiPolygon srsName="EPSG:2154">
			<polygonMember>
			<Polygon>
			<outerBoundaryIs>
			<LinearRing>
			<coordinates cs="," ts=" ">767666.07183931,6264670.2452116 767612.29116582,6264662.30291805 767606.50463766,6264642.44718417 767565.09124986,6264606.02609517 767537.97399044,6264576.52614769 767510.97019236,6264551.33773099 767430.1857208,6264543.50889878 767375.15697262,6264562.34348063 767386.73002894,6264589.12035603 767423.71842462,6264621.45683692 767532.18746228,6264720.84896766 767542.62590524,6264785.86231346 767571.33162336,6264842.25259768 767644.40072404,6264783.70654806 767654.04493764,6264745.35661634 767676.51028226,6264722.55088771 767666.07183931,6264670.2452116
			</coordinates>
			</LinearRing>
			</outerBoundaryIs>
			</Polygon>
			</polygonMember>
			</MultiPolygon>
		</qgs:geometry>
		<qgs:utilisateur>ornitho_ll</qgs:utilisateur>
		<qgs:id_site>PIERREBLANCHE_4</qgs:id_site>
		<qgs:reproduction>f</qgs:reproduction>
		<qgs:date_maj/>
		<qgs:toponyme>Pierre Blanche Aresquiers</qgs:toponyme>
		<qgs:gid>446</qgs:gid>
	</qgs:Sites_de_repro>
</gml:featureMember>
```

**Comment by rldhont** [comment]
Hi @mathieubossaert If the `mdo.affiche_repro_site` is a **view**, QGIS build it's own FeatureId based on QGIS internal index of the first read of the feature. FeatureId is a mess!

**Comment by mathieubossaert** [comment]
So is it a regression of QGIS ? Because it works fine with 2.14... and it's not so rare to use views and rules or triggers to edit data.

**Comment by gioman** [comment]
> So is it a regression of QGIS ? Because it works fine with 2.14.

@mathieubossaert unfortunately not the first one affecting qgis server 2.18 compared to 2.14

**Comment by mathieubossaert** [comment]
OK I think it could explain a bug we have with QGIS Desktop 2.18.

I expose a spatial view to my colleagues. When I select some features on the map, the good feautures are selected into the attribute table. When I ask for a zoom to selected from the attribute table, zommed feature is a wrong one.
Problem is "solved" by replacing the view by a materialized view...

**Comment by nboisteault** [comment]
Does issue still occur with QGIS Server 3.4 and LWC 3.3 ?

**Comment by rldhont** [comment]
It has been fixed in QGIS 3.4 and upper.

It has to be added to documentation that the prmary key has to be well defined.
